lib/Transforms/ForwardStoreToLoad/ForwardStoreToLoad.cpp

```diff
@@ -150,6 +150,82 @@ LogicalResult ForwardSingleStoreToLoad::matchAndRewrite(
   return failure();
 }
 
+bool RemoveUnusedStore::isPostDominated(Operation *potentialOp,
+                                        memref::StoreOp &storeOp) const {
+  if (!dominanceInfo.properlyDominates(storeOp.getOperation(), potentialOp)) {
+    LLVM_DEBUG(llvm::dbgs()
+               << "store of is not properly dominated by potential store\n");
+    return false;
+  }
+
+  // Probably want to relax this at some point in the future.
+  if (storeOp->getBlock() != potentialOp->getBlock()) {
+    LLVM_DEBUG(llvm::dbgs() << "store ops are not in the same block\n");
+    return false;
+  }
+
+  return llvm::TypeSwitch<Operation &, bool>(*potentialOp)
+      .Case<memref::StoreOp>([&](auto potentialStore) {
+        ValueRange storeIndices = storeOp.getIndices();
+        ValueRange potentialStoreIndices = potentialStore.getIndices();
+        if (storeIndices != potentialStoreIndices) {
+          LLVM_DEBUG(llvm::dbgs()
+                     << "store ops do not have matching indices\n");
+          return false;
+        }
+
+        // Naively scan through the operations between the two ops and check if
+        // a read prevents store removal.
+        for (auto currentNode = storeOp->getNextNode();
+             currentNode != potentialStore.getOperation();
+             currentNode = currentNode->getNextNode()) {
+          if (currentNode->getNumRegions() > 0) {
+            // Op can have control flow
+            LLVM_DEBUG(llvm::dbgs() << "an op with control flow is between the "
+                                       "store ops\n");
+            return false;
+          }
+          if (auto op = dyn_cast<affine::AffineLoadOp>(currentNode)) {
+            // If we encounter an affine load op then fail conservatively. This
+            // pass should have already run AffineLoadLowering to convert all
+            // possible affine loads to memref loads.
+            LLVM_DEBUG(llvm::dbgs() << "an intermediate load op was found\n");
+            return false;
+          }
```

Comment:
How about adding a test case for this? Would be useful to catch if a PR accidentally changes the pass ordering.

Reply:
The pattern for converting to memref loads is contained within this pass btw, so it's not dependent on external callers to this pass.

This means adding a testcase that still has an affine.load at this stage pretty hard haha. But I'll add a testcase with affine.loads - it won't exercise this branch though. But if it fails, then it would be exercising this path.